pkg/flow/user_component.go

```diff
@@ -0,0 +1,284 @@
+package flow
+
+import (
+	"fmt"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/rfratto/gohcl"
+)
+
+// userComponentName is the fully-qualified name of a userComponent.
+type userComponentName []string
+
+// blockToComponentName gets the userComponentName from an HCL block.
+func blockToComponentName(b *hcl.Block) userComponentName {
+	name := make(userComponentName, 0, 1+len(b.Labels))
+	name = append(name, b.Type)
+	name = append(name, b.Labels...)
+	return name
+}
+
+func (r userComponentName) String() string {
+	return strings.Join(r, ".")
+}
+
+func (r userComponentName) Equals(other userComponentName) bool {
+	if len(r) != len(other) {
+		return false
+	}
+	for i := 0; i < len(r); i++ {
+		if r[i] != other[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// userComponentOptions are options shared between all components.
+type userComponentOptions struct {
+	Logger        log.Logger              // Base logger shared between components
+	DataPath      string                  // Path where components may store data
+	OnStateChange func(uc *userComponent) // Invoked to propagate state change
```

Comment:
Yeah, we need a way to tell the controller specifically which component got updated. Using a shared method with an argument for the component instead of getting the component from a closure made it easier to build.

Reply:
(I tried the closure approach at first but you had to lazily set the function and update its implementation after creating the component, which wasn't pretty)